gcp/api/cursor.py

```diff
@@ -0,0 +1,121 @@
+from typing import Self
+import typing
+
+from google.cloud import ndb
+import google.cloud.ndb.exceptions as ndb_exceptions
+
+_FIRST_PAGE_TOKEN = base64.urlsafe_b64encode(b'FIRST_PAGE_TOKEN').decode()
+# Use ':' as the separator as it doesn't appear in urlsafe_b64encode
+# (which is what is used for both _FIRST_PAGE_TOKEN, and ndb.Cursor.urlsafe())
+_METADATA_SEPARATOR = ':'
+
+
+class _QueryCursorState(Enum):
+  """
+  Stores the 3 states a query cursor can be in.
+  """
+
+  # The cursor has reached the end, no need to return to the user
+  ENDED = 0
+  # The cursor is at the very start of the query, no cursor needs to be
+  # set when querying ndb
+  STARTED = 1
+  # ndb.Cursor is set and in progress
+  IN_PROGRESS = 2
+
+
+class QueryCursor:
+  """
+  Custom cursor class that wraps the ndb cursor.
+  Allows us to represent the "starting" cursor.
+
+  The default state is ENDED with no ndb.Cursor.
+
+  This type could have 3 states encoded in _QueryCursorState.
+  If the current state is IN_PROGRESS, self.cursor will not be None.
+  """
+
+  cursor: ndb.Cursor | None = None
+  cursor_state: _QueryCursorState = _QueryCursorState.ENDED
+  query_number: int = 1
+
+  @classmethod
+  def from_page_token(cls, page_token: str | None) -> Self:
+    """Generate a query cursor from a url safe page token."""
+    qc = cls()
+
+    if not page_token:
+      qc.cursor_state = _QueryCursorState.STARTED
+      qc.cursor = None
+      return qc
+
+    split_values = page_token.split(_METADATA_SEPARATOR, 1)
+    if len(split_values) == 2:
+      page_token = split_values[1]
+      try:
+        qc.query_number = int(split_values[0])
+      except ValueError as e:
+        raise ValueError('Invalid page token.') from e
+
+    if not page_token or page_token == _FIRST_PAGE_TOKEN:
+      qc.cursor_state = _QueryCursorState.STARTED
+      qc.cursor = None
+      return qc
+
+    qc.cursor = ndb.Cursor(urlsafe=page_token)
+    qc.cursor_state = _QueryCursorState.IN_PROGRESS
+    return qc
+
+  def update_from_iterator(self, it: ndb.QueryIterator) -> None:
+    try:
+      self.cursor = typing.cast(ndb.Cursor, it.cursor_after())
+      self.cursor_state = _QueryCursorState.IN_PROGRESS
+    except ndb_exceptions.BadArgumentError:
+      self.cursor = None
+      self.cursor_state = _QueryCursorState.STARTED
+
+  def get_cursor(self) -> ndb.Cursor | None:
+    if self.cursor_state == _QueryCursorState.IN_PROGRESS:
+      return self.cursor
+
+    return None
+
+  def ended(self) -> bool:
+    return self.cursor_state == _QueryCursorState.ENDED
+
+  def url_safe_encode(self) -> str | None:
+    """Create a url safe page token to pass back to the API caller"""
+    cursor_part: str = ''
+    match self.cursor_state:
+      case _QueryCursorState.STARTED:
+        cursor_part = _FIRST_PAGE_TOKEN
+      case _QueryCursorState.IN_PROGRESS:
+        # Assume that IN_PROGRESS means self.cursor is always set.
+        # Loudly throw an exception if this is not the case
```

Comment:
If I understand this comment correctly, you're relying on `self.cursor.urlsafe().decode()` to raise an exception? Can you please expand the docstring to mention this? https://google.github.io/styleguide/pyguide.html#s3.8.3-functions-and-methods

Reply:
It will  raise an ArgumentError since urlsafe() will not exist on None. I'm not sure if I should be documenting this in the docstring since it's an implementation/programmer error of QueryCursor if that is ever the case. This is not something the caller should be accounting for, as it should never happen.

> You should not document exceptions that get raised if the API specified in the docstring is violated (because this would paradoxically make behavior under violation of the API part of the API).

Reply:
https://www.xkcd.com/2200/😆